src/test/ui/type-mismatch.stderr

```diff
@@ -0,0 +1,432 @@
+  --> $DIR/type-mismatch.rs:69:26
+   |
+LL |     want::<&Foo<foo, B>>(f); //~ ERROR mismatched types
+   |                          ^
+   |                          |
+   |                          expected reference, found struct `Foo`
+   |                          help: consider borrowing here: `&f`
+   |
+   = note: expected type `&Foo<foo, B>`
+              found type `Foo<foo, B>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:73:19
+   |
+LL |     want::<usize>(f); //~ ERROR mismatched types
+   |                   ^ expected usize, found struct `Foo`
+   |
+   = note: expected type `usize`
+              found type `Foo<foo, B, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:74:17
+   |
+LL |     want::<foo>(f); //~ ERROR mismatched types
+   |                 ^ expected struct `foo`, found struct `Foo`
+   |
+   = note: expected type `foo`
+              found type `Foo<foo, B, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:75:17
+   |
+LL |     want::<bar>(f); //~ ERROR mismatched types
+   |                 ^ expected struct `bar`, found struct `Foo`
+   |
+   = note: expected type `bar`
+              found type `Foo<foo, B, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:76:24
+   |
+LL |     want::<Foo<usize>>(f); //~ ERROR mismatched types
+   |                        ^ expected usize, found struct `foo`
+   |
+   = note: expected type `Foo<usize, A, B>`
+              found type `Foo<foo, B, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:77:27
+   |
+LL |     want::<Foo<usize, B>>(f); //~ ERROR mismatched types
+   |                           ^ expected usize, found struct `foo`
+   |
+   = note: expected type `Foo<usize, _, B>`
+              found type `Foo<foo, _, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:78:22
+   |
+LL |     want::<Foo<foo>>(f); //~ ERROR mismatched types
+   |                      ^ expected struct `A`, found struct `B`
+   |
+   = note: expected type `Foo<_, A, B>`
+              found type `Foo<_, B, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:79:25
+   |
+LL |     want::<Foo<foo, B>>(f); //~ ERROR mismatched types
+   |                         ^ expected struct `B`, found struct `A`
+   |
+   = note: expected type `Foo<_, _, B>`
+              found type `Foo<_, _, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:80:22
+   |
+LL |     want::<Foo<bar>>(f); //~ ERROR mismatched types
+   |                      ^ expected struct `bar`, found struct `foo`
+   |
+   = note: expected type `Foo<bar, A, B>`
+              found type `Foo<foo, B, A>`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch.rs:81:25
+   |
+LL |     want::<Foo<bar, B>>(f); //~ ERROR mismatched types
+   |                         ^ expected struct `bar`, found struct `foo`
+   |
+   = note: expected type `Foo<bar, _, B>`
+              found type `Foo<foo, _, A>`
```

Comment:
Just out of curiosity, are `bar`, `foo`, `B` and `A` highlighted in the output here? I would think yes, but want to confirm (regardless, this PR is going in, I'm asking just to see if I have to file a ticket to add that).

Reply:
Yes, they are. It would be great if the ui tests actually tested for that too. It would become more apparent that there's some inconsistency for e.g. `usize`/`Foo<usize>` where everything is highlighted as opposed to `foo`/`Foo<foo>` where `foo` is not highlighted, which is because usize is not an Adt and doesn't fall in the same code path as foo.

Reply:
I think that the latter is not because of that but rather because `foo` is detected as the first type argument to `Foo`, which means that the think to highlight is only `Foo`, indirectly telling you "you need to wrap this with a `Foo`!". It's the same as when you have an `Option<Foo>` and a `Result<Foo, _>` was expected, `Foo` doesn't get highlighted.

> It would be great if the ui tests actually tested for that too.

It would, but it probably would be a pain to look at :-/ The current situation is _good enough™_ that we haven't pushed to change it yet.

Reply:
No, really, it's only because `foo` is an Adt, and `usize` isn't. The comparison between `foo` and `Foo<foo>` hits this pattern:
https://github.com/rust-lang/rust/blob/64f7de92166f1f7d940575feff57ac33bc384550/src/librustc/infer/error_reporting/mod.rs#L715

while the comparison between `usize` and `Foo<usize>` hits this one:
https://github.com/rust-lang/rust/blob/64f7de92166f1f7d940575feff57ac33bc384550/src/librustc/infer/error_reporting/mod.rs#L857

Reply:
Oh! You're absolutely right. I though I had expanded a few more branches to account for this case before, but now looking at the code in master I see that I must have never submitted that PR.